ml4gw/dataloading/chunked_dataset.py

```diff
@@ -0,0 +1,280 @@
+
+        num_chunks = (self.chunks_per_epoch - 1) // self.num_workers + 1
+        return num_chunks * self.num_workers * self.batches_per_chunk
+
+    def iter_epoch(self):
+        # slice kernels out a flattened chunk tensor
+        # index-for-index. We'll account for batch/
+        # channel indices by introducing offsets later on
+        idx = torch.arange(self.kernel_size, device=self.device)
+        idx = idx.view(1, 1, -1)
+        idx = idx.repeat(self.batch_size, self.num_channels, 1)
+
+        # this will just be a set of aranged channel indices
+        # repeated to offset the kernel indices in the
+        # flattened chunk tensor
+        channel_idx = torch.arange(self.num_channels, device=self.device)
+        channel_idx = channel_idx.view(1, -1, 1)
+        channel_idx = channel_idx.repeat(self.batch_size, 1, self.kernel_size)
+        idx += channel_idx * self.chunk_size
+
+        for chunk in self.chunk_loader:
+            # record the number of rows in the chunk, then
+            # flatten it to make it easier to slice
+            num_chunks, _, chunk_size = chunk.shape
+            chunk = chunk.to(self.device).reshape(-1)
+
+            # generate batches from the current chunk
+            for _ in range(self.batches_per_chunk):
+                # if we're sampling coincidentally, we only need
+                # to sample indices on a per-batch-element basis.
+                # Otherwise, we'll need indices for both each
+                # batch sample _and_ each channel with each sample
+                if self.coincident:
+                    size = (self.batch_size,)
+                else:
+                    size = (self.batch_size, self.num_channels)
+
+                # first sample the indices of which chunk elements
+                # we're going to read batch elements from
+                chunk_idx = torch.randint(
+                    0, num_chunks, size=size, device=self.device
+                )
+
+                # account for the offset this batch element
+                # introduces in the flattened array
+                chunk_idx *= self.num_channels * self.chunk_size
+                chunk_idx = chunk_idx.view(self.batch_size, -1, 1)
+                chunk_idx = chunk_idx + idx
+
+                # now sample the start index within each chunk
+                # element we're going to grab our time windows from
+                time_idx = torch.randint(
+                    0,
+                    chunk_size - self.kernel_size,
+                    size=size,
+                    device=self.device,
+                )
+                time_idx = time_idx.view(self.batch_size, -1, 1)
+
+                # there's no additional offset factor to account for here
+                chunk_idx += time_idx
+
+                # now slice this 3D tensor from our flattened chunk
+                yield chunk[chunk_idx]
+
+    def __iter__(self):
+        return self.iter_epoch()
```

Comment:
Question for my own sake: why make a separate function if all `__iter__` does is call it?

Reply:
This is a great question. The one real requirement of `__iter__` is that it return an object with a `__next__` method. That's what's functionally happening under the hood when you do e.g. `for i in iterable`: Python calls the `__iter__` method on `iterable`, then at each step in the loop assigns to `i` the output of the `__next__` method of the object that gets returned by `__iter__` (the `iterator`, not necessarily `iterable` _itself_). This repeats until calling `__next__` raises a `StopIteration`.

So an `Iterable` is an object with an `__iter__` method that shouldn't do the iteration itself, but rather return an `Iterator`, an object with a `__next__` method. Some objects will act as both `Iterable` and `Iterator`, using `__iter__` to initialize some iteration parameters and then `return self`, so that its own `__next__` method is what gets called at each loop. Each call to `__next__` will typically increment those parameters. Take for example the following dummy implementation of `range`

```python
class Range:
    def __init__(self, start, stop, step):
        self.start = start
        self.stop = stop
        self.step = step
        self._i = None

    def __iter__(self):
        self._i = self.start
        return self

    def __next__(self):
        if self._i >= self.stop:
            self._i = None
            raise StopIteration
        value = self._i + 0
        self._i = self._i + self.step
        return value
```
This is fine, but as is noted in [this](https://dbader.org/blog/python-iterators) great overview: "Python iterators normally can’t be 'reset'—once they’re exhausted they’re supposed to raise `StopIteration` every time `next()` is called on them." So a common pattern is to break up the `Iterable` and the `Iterator` into separate objects, such that every time you call `iter` on the `Iterable`, it returns a fresh `Iterator` that you can loop through.

So all I'm doing here is rather than

- creating a dedicated `Iterator` object that would have to borrow a bunch of attributes from this object
- creating a `__next__` method where I have to keep track of incrementing indices and raising `StopIteration` myself, and technically violating this Python rule-of-thumb (since we'll iterate through this dataset several times)

I create a generator using an internal method, and return that as my `Iterator`.

Reply:
Got it, okay, thanks for explaining all that.